File: src/main/java/me/rsman/BetterMinecraftCore/formatters/LoreFormatter.java
package me.rsman.BetterMinecraftCore.formatters;

import me.rsman.BetterMinecraftCore.BetterMinecraftCore;
import me.rsman.BetterMinecraftCore.Managers.EnchantManager;
import me.rsman.BetterMinecraftCore.Managers.ItemManager;
import me.rsman.BetterMinecraftCore.configs.containers.AttributeLangContainer;
import me.rsman.BetterMinecraftCore.configs.containers.EnchantLangContainer;
import me.rsman.BetterMinecraftCore.configs.containers.GlobalConfigContainer;
import me.rsman.BetterMinecraftCore.enums.EAttributes;
import me.rsman.BetterMinecraftCore.enums.EEnchants;
import me.rsman.BetterMinecraftCore.utils.NBT;
import me.rsman.BetterMinecraftCore.utils.RomanNumber;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataType;

import java.util.*;

public final class LoreFormatter {
    public static List<String> format(ItemStack item){

        List<String> lore = new ArrayList<>(formatAttributesLore(item));

        lore.addAll(formatEnchantsLore(item));

        //custom lore
        String CustomLore = (String) NBT.get(item, "lore", PersistentDataType.STRING);
        if(CustomLore != null && !CustomLore.equals("")) {
            lore.add("");
            CustomLore = CustomLore.replace("&", "§");
            lore.addAll(Arrays.asList(CustomLore.split("\\|")));
        }

        //unbreakable lore
        if(item.hasItemMeta() && Objects.requireNonNull(item.getItemMeta()).isUnbreakable()){
            lore.add("");
            lore.add("§cUnbreakable");
        }
        return lore;
    }

    public static List<String> formatAttributesLore(ItemStack item){
        List<String> attributesLore = new ArrayList<>();
        for (String attr : EAttributes.getAllKeys()) {
            long attrVal = ItemManager.getItemAttr(item, attr);
            long attrModifierVal = ItemManager.getItemEnchantAttr(item, attr);
            if(attrVal != 0 || attrModifierVal != 0){
                boolean percent = EAttributes.getAllPercentKeys().contains(attr);
                String attrTrans = AttributeLangContainer.getInstance().getTranslation(attr);
                if(attrTrans == null) attrTrans = attr.substring(0, 1).toUpperCase() + attr.substring(1);
                attributesLore.add(
                        GlobalConfigContainer.getInstance().getAttribute_display_format().replace("&","§")
                        .replace("{attr_name}", attrTrans.replace("&", "§"))
                        .replace("{value}", attrVal + (percent ? "%" : "")) +
                            (attrModifierVal >0 ?
                                GlobalConfigContainer.getInstance().getAttribute_modifier_display_format().replace("&","§")
                                    .replace("{value}", attrModifierVal + (percent ? "%" : ""))
                            : "")
                );
            }
        }
        return attributesLore;
    }

    public static List<String> formatEnchantsLore(ItemStack item){
        List<String> enchantsLore = new ArrayList<>();
        boolean first = true;
        StringBuilder line = new StringBuilder("§7");
        for (Map.Entry<Enchantment, Integer> enchant: item.getEnchantments().entrySet()) {
            if(enchant.getValue() <= 0)continue;
            if(first){ first = false; enchantsLore.add(""); }

            String enchantTrans = EnchantLangContainer.getInstance().getTranslation(
                    EEnchants.getEnumKeyFromKey(
                            enchant.getKey().getKey().toString().replaceFirst("minecraft:", "")
                    )
            );
            if(enchantTrans == null) enchantTrans = enchant.getKey().getKey().toString();
            String enchantName = GlobalConfigContainer.getInstance().getEnchant_display_format().replace("&","§")
                    .replace("{ench_name}", enchantTrans.replace("&", "§"))
                    .replace("{value}", RomanNumber.toRoman(enchant.getValue()))
                    .replace("{value_int}", enchant.getValue().toString());

            if(line.length() < 30){
                if(!line.toString().equals("§7")) line.append(GlobalConfigContainer.getInstance().getEnchant_separator_display_format().replace("&", "§")).append(" ");
                line.append(enchantName);
            } else {
                enchantsLore.add(line + GlobalConfigContainer.getInstance().getEnchant_separator_display_format().replace("&", "§"));
                line = new StringBuilder("§7" + enchantName);
            }
        }
        if(!line.toString().equals("§7")){
            enchantsLore.add(line.toString());
        }
        return enchantsLore;
    }

}

File: src/main/kotlin/me/rsman/BetterMinecraftCore/formatters/LoreFormatter.kt
package me.rsman.BetterMinecraftCore.formatters

import me.rsman.BetterMinecraftCore.enums.EEnchants
import me.rsman.BetterMinecraftCore.enums.EAttributes
import me.rsman.BetterMinecraftCore.configs.containers.GlobalConfigContainer
import org.bukkit.inventory.ItemStack
import me.rsman.BetterMinecraftCore.configs.containers.AttributeLangContainer
import java.lang.StringBuilder
import me.rsman.BetterMinecraftCore.configs.containers.EnchantLangContainer
import me.rsman.BetterMinecraftCore.extensions.customLore
import me.rsman.BetterMinecraftCore.extensions.getEnchAtribute
import me.rsman.BetterMinecraftCore.extensions.getStoredAtribute
import me.rsman.BetterMinecraftCore.extensions.isUnbreakable
import me.rsman.BetterMinecraftCore.utils.romanString
import java.util.*

object LoreFormatter {
    fun format(item: ItemStack): List<String> {
        val lore = formatAttributesLore(item).toMutableList()
        lore.addAll(formatEnchantsLore(item))

        //custom lore
        lore.addAll(item.customLore?.map { it.replace("&", "§") } ?: listOf())

        //unbreakable lore
        if (item.isUnbreakable) {
            lore.add("")
            lore.add("§cUnbreakable")
        }
        return lore.map { PapiManager.parseText(null, it) }.toMutableList()
    }

    fun formatAttributesLore(item: ItemStack): List<String> {
        val attributesLore: MutableList<String> = ArrayList()
        for (attr in EAttributes.values()) {
            val attrVal = item.getStoredAtribute(attr)
            val attrModifierVal = item.getEnchAtribute(attr)
            if (attrVal != 0L || attrModifierVal != 0L) {
                val attrTrans = AttributeLangContainer.instance!!.getTranslation(attr.key) ?: attr.key.replaceFirstChar { it.uppercase() }
                attributesLore.add(
                        GlobalConfigContainer.instance?.attribute_display_format?.replace("&", "§")
                                ?.replace("{attr_name}", attrTrans.replace("&", "§"))
                                ?.replace("{value}", attrVal.toString() + if (attr.isPercent) "%" else "") +
                                if (attrModifierVal > 0) GlobalConfigContainer.instance?.attribute_modifier_display_format?.replace("&", "§")
                                        ?.replace("{value}", attrModifierVal.toString() + if (attr.isPercent) "%" else "") else ""
                )
            }
        }
        return attributesLore
    }

    fun formatEnchantsLore(item: ItemStack): List<String> {
        val enchantsLore: MutableList<String> = ArrayList()
        var first = true
        var line = StringBuilder("§7")
        for ((key, value) in item.enchantments) {
            if (value <= 0) continue
            if (first) {
                first = false
                enchantsLore.add("")
            }
            val enchantTrans = EnchantLangContainer.instance?.getTranslation(
                    EEnchants.fromKey(
                            key.key.toString().replaceFirst("minecraft:".toRegex(), "")
                    )?.key
            ) ?: key.key.toString()
            val enchantName = GlobalConfigContainer.instance!!.enchant_display_format!!.replace("&", "§")
                    .replace("{ench_name}", enchantTrans.replace("&", "§"))
                    .replace("{value}", value.romanString!!)
                    .replace("{value_int}", value.toString())
            if (line.length < 30) {
                if (line.toString() != "§7") line.append(GlobalConfigContainer.instance!!.enchant_separator_display_format!!.replace("&", "§")).append(" ")
                line.append(enchantName)
            } else {
                enchantsLore.add(line.toString() + GlobalConfigContainer.instance!!.enchant_separator_display_format!!.replace("&", "§"))
                line = StringBuilder("§7$enchantName")
            }
        }
        if (line.toString() != "§7") {
            enchantsLore.add(line.toString())
        }
        return enchantsLore
    }
}